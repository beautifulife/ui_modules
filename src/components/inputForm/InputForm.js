import React, { Component } from 'react';
import PropTypes from 'prop-types';
import './InputForm.scss';

class InputForm extends Component {
  constructor(props) {
    super(props);
    const { initialValue } = this.props;
    this.state = {
      inputValue: initialValue
    };

    this.handleChange = this.handleChange.bind(this);
    this.handleClick = this.handleClick.bind(this);
  }

  handleChange(event) {
    if (event.currentTarget.value.length <= 500) {
      this.setState({
        inputValue: event.currentTarget.value
      });
    }
  }

  handleClick() {
    const { inputValue } = this.state;
    const { onSaveClick } = this.props;

    onSaveClick(inputValue);
  }

  render() {
    const { inputValue } = this.state;
    const { initialValue, uiState } = this.props;

    const isInputValueChanged = inputValue !== initialValue;
    const isInputDisabled = uiState === 'disabled' || uiState === 'readonly';
    const wordCountToLimit = 500 - inputValue.length;

    return (
      <div className="InputForm">
        <div
          className={`InputForm__wrapper ${
            inputValue && isInputValueChanged ? 'active' : ''
          }`}
        >
          <textarea
            className={`InputForm__textarea ${
              isInputDisabled ? 'disabled' : ''
            }`}
            value={inputValue}
            placeholder="입력창 입니다."
            onChange={this.handleChange}
            disabled={isInputDisabled}
          />
          <span className="InputForm__word_count">{wordCountToLimit}</span>
        </div>
        {inputValue && isInputValueChanged && (
          <button
            className="InputForm__save-btn"
            type="submit"
            onClick={this.handleClick}
          >
            Save
          </button>
        )}
      </div>
    );
  }
}

InputForm.propTypes = {
  initialValue: PropTypes.string,
  uiState: PropTypes.string,
  onSaveClick: PropTypes.func.isRequired
};

InputForm.defaultProps = {
  initialValue: '',
  uiState: 'default'
};

export default InputForm;
