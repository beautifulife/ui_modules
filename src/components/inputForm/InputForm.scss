.InputForm {
  display: table;
  width: 600px;
  margin-bottom: 5px;

  &__cell {
    display: table-cell;
    position: relative;
  }

  &__textarea {
    display: block;
    width: 100%;
    height: 50px;
    padding: 5px;
    border: 1px solid #eee;
    resize: none;

    &:focus {
      outline: 1px solid #31c089;
    }

    &.disabled {
      background-color: #f5f5f5;
      opacity: 0.7;
    }
  }

  &__word_count {
    position: absolute;
    right: -5px;
    bottom: 5px;
    font-size: 0.6rem;
    color: rgba(0, 0, 0, 0.45);
  }

  &__save-btn {
    display: block;
    width: 50px;
    height: 50px;
    border: 1px solid #eee;
    color: #31c089;
    background: none;
    cursor: pointer;
  }
}
