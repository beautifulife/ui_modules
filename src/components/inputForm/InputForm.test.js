import React from 'react';
import { shallow } from 'enzyme';
import InputForm from './InputForm';

describe('InputForm', () => {
  describe('about render', () => {
    const wrapper = shallow(<InputForm />);

    it('compare with snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
