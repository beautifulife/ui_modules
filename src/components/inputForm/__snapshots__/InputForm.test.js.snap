// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputForm about render compare with snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <InputForm
    initialValue=""
    uiState="default"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="InputForm__wrapper "
        >
          <textarea
            className="InputForm__textarea "
            disabled={false}
            onChange={[Function]}
            placeholder="입력창 입니다."
            value=""
          />
          <span
            className="InputForm__word_count"
          >
            500
          </span>
        </div>,
        "",
      ],
      "className": "InputForm",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <textarea
              className="InputForm__textarea "
              disabled={false}
              onChange={[Function]}
              placeholder="입력창 입니다."
              value=""
            />,
            <span
              className="InputForm__word_count"
            >
              500
            </span>,
          ],
          "className": "InputForm__wrapper ",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "InputForm__textarea ",
              "disabled": false,
              "onChange": [Function],
              "placeholder": "입력창 입니다.",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "textarea",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": 500,
              "className": "InputForm__word_count",
            },
            "ref": null,
            "rendered": 500,
            "type": "span",
          },
        ],
        "type": "div",
      },
      "",
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="InputForm__wrapper "
          >
            <textarea
              className="InputForm__textarea "
              disabled={false}
              onChange={[Function]}
              placeholder="입력창 입니다."
              value=""
            />
            <span
              className="InputForm__word_count"
            >
              500
            </span>
          </div>,
          "",
        ],
        "className": "InputForm",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <textarea
                className="InputForm__textarea "
                disabled={false}
                onChange={[Function]}
                placeholder="입력창 입니다."
                value=""
              />,
              <span
                className="InputForm__word_count"
              >
                500
              </span>,
            ],
            "className": "InputForm__wrapper ",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "InputForm__textarea ",
                "disabled": false,
                "onChange": [Function],
                "placeholder": "입력창 입니다.",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "textarea",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": 500,
                "className": "InputForm__word_count",
              },
              "ref": null,
              "rendered": 500,
              "type": "span",
            },
          ],
          "type": "div",
        },
        "",
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
